src/core/Akka.Streams/Dsl/RepeatPrevious.cs

```diff
@@ -0,0 +1,105 @@
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using Akka.Streams.Stage;
+using Akka.Util;
+
+namespace Akka.Streams.Dsl
+{
+    /// <summary>
+    /// Delegate used to customize replacement functions.
+    ///
+    /// This is used for things like calling <see cref="IDisposable.Dispose"/> on the previous element.
+    /// when a new value is emitted.
+    /// </summary>
+    /// <typeparam name="T">The type of element handled by the <see cref="RepeatPrevious{T}"/></typeparam>
+    public delegate void SwapPrevious<T>(T previousValue, T newValue);
+
+    /// <summary>
+    /// Repeats the previous element from upstream until it's replaced by a new value.
+    ///
+    /// This is designed to allow fan-in stages where output from one of the sources is intermittent / infrequent
+    /// and users just want the previous value to be reused.
+    /// </summary>
+    /// <typeparam name="T">The output type.</typeparam>
+    public sealed class RepeatPrevious<T> : GraphStage<FlowShape<T, T>>
+    {
+        private readonly Inlet<T> _in = new Inlet<T>("RepeatPrevious.in");
+        private readonly Outlet<T> _out = new Outlet<T>("RepeatPrevious.out");
+
+        public override FlowShape<T, T> Shape => new FlowShape<T, T>(_in, _out);
+        private readonly SwapPrevious<T> _swapPrevious;
+
+        /// <summary>
+        /// Do nothing by default
+        /// </summary>
+        private static readonly SwapPrevious<T> DefaultSwap = (value, newValue) => { };
+
+        public RepeatPrevious() : this(DefaultSwap)
+        {
+        }
+
+        public RepeatPrevious(SwapPrevious<T> swapPrevious)
+        {
+            _swapPrevious = swapPrevious;
+        }
+
+        protected override GraphStageLogic CreateLogic(Attributes inheritedAttributes) =>
+            new Logic(this, _swapPrevious);
+
+        private sealed class Logic : InAndOutGraphStageLogic
+        {
+            private readonly RepeatPrevious<T> _stage;
+            private Option<T> _last;
+            private readonly SwapPrevious<T> _swapPrevious;
+
+            public Logic(RepeatPrevious<T> stage, SwapPrevious<T> swapPrevious) : base(stage.Shape)
+            {
+                _stage = stage;
+                _swapPrevious = swapPrevious;
+
+                SetHandler(_stage._in, this);
+                SetHandler(_stage._out, this);
+            }
+
+            public override void OnPush()
+            {
+                var next = Grab(_stage._in);
+                if (_last.HasValue)
+                    _swapPrevious(_last.Value, next);
+                _last = next;
+
+                if (IsAvailable(_stage._out))
+                {
+                    Push(_stage._out, _last.Value);
+                }
+            }
+
+            public override void OnPull()
+            {
+                if (_last.HasValue)
+                {
+                    if (!HasBeenPulled(_stage._in))
```

Comment:
To your comments about repeat elements in the test, I wonder if guard code in pull would help. Expand does something like this:

```
if (_iterator.HasNext())
                {
                    if (!_expanded)
                    {
                        _expanded = true;
                        if (IsClosed(_stage.In))
                        {
                            Push(_stage.Out, _iterator.Next());
                            CompleteStage();
                        }
                        else
                        {
                            // expand needs to pull first to be "fair" when upstream is not actually slow
                            Pull(_stage.In);
                            Push(_stage.Out, _iterator.Next());
                        }
                    }
                    else
                        Push(_stage.Out, _iterator.Next());
                }
```

And I wonder if that IsClosed guard would help?

Reply:
I'll take a look at this! I tried coming up with something similar to this but I just kept moving the problem somewhere else in the test suite. Let me give this a look though.

Reply:
Yeah I tried this again and it's not going to work - the issue is that in order to detect that the upstream is closed it has to be pulled a second time and during that timeframe we're going to emit an extra item.

The `Expand` specs don't really check for the case where the upstream terminates - they're mostly just testing that expansion doesn't happen so long as downstream can keep up with demand.